File: leetcode/493_Reverse_Pairs/493_Reverse_Pairs.java
class Solution {
    private int[] tree;
    private int m;
    private void build(int[] nums) {
        for (m = 1; m <= nums.length; m <<= 1);
        tree = new int[2 * m];
        for (int i = m + 1; i <= m + nums.length; i++) tree[i] = 0;
    }
    private void update(int index) {
        int pos = index + m + 1;
        tree[pos] += 1;
        for (pos >>= 1; pos > 0; pos >>= 1) tree[pos] = tree[pos << 1] + tree[pos << 1 | 1];
    }
    private int query(int s, int e) {
        int res = 0;
        for (s += m, e += m + 2; (s^e^1) != 0; s >>= 1, e >>= 1) {
            if (s % 2 == 0) res += tree[s + 1];
            if (e % 2 == 1) res += tree[e - 1];
        }
        return res;
    }
    public int reversePairs(int[] nums) {
        build(nums);
        long[] numsSorted = new long[nums.length];
        for (int i = 0; i < nums.length; i++) numsSorted[i] = nums[i];
        Arrays.sort(numsSorted);
        int res = 0;
        for (int i = 0; i < nums.length; i++) {
            int pos = Arrays.binarySearch(numsSorted, nums[i] * 2L + 1L);
            if (pos < 0) pos = -1 * (pos + 1);
            res += query(pos, nums.length - 1);
            update(Arrays.binarySearch(numsSorted, nums[i]));
        }
        return res;
    }
}

File: leetcode/493_Reverse_Pairs/493_Reverse_Pairs.cpp
class Solution {
public:
    class Node {
        public:
            int start;
            int end;
            int val;
            Node* left;
            Node* right;
            Node(int s, int e, int v): start(s), end(e), val(v) {}
    };
    Node* build(int start, int end) {
        Node* cur = new Node(start, end, 0);
        if (start != end) {
            int mid = start + (end - start) / 2;
            cur->left = build(start, mid);
            cur->right = build(mid + 1, end);
        }
        return cur;
    }
    void update(Node* n, int index) {
        if (n->start == n->end && n->start == index) n->val += 1;
        else {
            int mid = n->start + (n->end - n->start) / 2;
            if (mid >= index) update(n->left, index);
            else update(n->right, index);
            n->val = n->left->val + n->right->val;
        }
    }
    int query(Node* n, int start, int end) {
        if (n->start <= end && start <= n->end && start <= end) {
            if (n->start >= start && n->end <= end) return n->val;
            else return query(n->left, start, end) + query(n->right, start, end);
        } else return 0;
    }
    
    int reversePairs(vector<int>& nums) {
        if (nums.empty()) return 0;
        Node* root = build(0, nums.size() - 1);
        vector<int> numsSorted = vector<int>(nums.begin(), nums.end());
        sort(numsSorted.begin(), numsSorted.end());
        int res = 0;
        for (int j = 0; j < nums.size(); j++) {
            res += query(root, upper_bound(numsSorted.begin(), numsSorted.end(), 2LL * nums[j]) - numsSorted.begin(), nums.size() - 1);
            update(root, lower_bound(numsSorted.begin(), numsSorted.end(), nums[j]) - numsSorted.begin());
        }
        return res;
    }
};
